src/dep.mli

```diff
@@ -1,29 +1,47 @@
 open Stdune
 
-type t
+type t = private
+  | Env of Env.Var.t
+  | File of Path.t
+  | Glob of Path.t * Path.t Predicate.t
+  | Universe
 
 val file : Path.t -> t
 val env : Env.Var.t -> t
 val universe : t
+val glob : dir:Path.t -> Path.t Predicate.t -> t
 
 val compare : t -> t -> Ordering.t
 
 val pp : t Fmt.t
 
+type eval_pred = dir:Path.t -> Path.t Predicate.t -> Path.Set.t
+
 module Set : sig
   include Set.S with type elt = t
 
   val has_universe : t -> bool
 
-  val paths : t -> Path.Set.t
+  val of_files : Path.t list -> t
+
+  val paths :  t -> eval_pred:eval_pred -> Path.Set.t
 
   val encode : t -> Dune_lang.t
 
-  val trace : t -> env:Env.t -> (string * Digest.t) list
+  val trace : t -> env:Env.t -> eval_pred:eval_pred -> (string * Digest.t) list
 
   val add_paths : t -> Path.Set.t -> t
 
-  val parallel_iter : t -> f:(Path.t -> unit Fiber.t) -> unit Fiber.t
+  val parallel_iter
```

Comment:
Why focus on `parallel_iter` here? That seems weird to me. Why not expose simply

```
val to_list : t -> elt list
val file_list : t -> Path.t list
```
and have the call site call `parallel_iter` if they need it?

Reply:
Eventually, we'd like to have an iterator over all the dependencies that does not allocate any intermediate lists. So I'm not a fan of making all these intermediate list conversions implicit.

Reply:
<sigh> I don't like that as a reason. 
Oh well at least there is one.

(I say I don't like it because I would much rather see something along the lines of `'a list_without_allocation = { iter : f:('a -> unit) -> unit }` as a tool to avoid allocation)

Reply:
Would that work with monadic iteration? We need unit Monad.t rather than just unit.

Reply:
Yeah, I guess so. Well, that's clearly a half-baked idea.